### ISSUE TITLE
Add debugging class

### ISSUE THREAD

**Issue by mindctrl** [restrictcontentpro/restrict-content-pro#1103]
Let's add an option to enable debugging to a custom log file. There's one in AffWP we can use.

**Comment by nosegraze** [comment]
PR: #1133

I only added one log so far so I could test it. To do:

- [ ] Add more logs.
- [ ] Replace uses of `WP_Logging`.
- [ ] Remove `WP_Logging` class.

One thing to consider: we talked about how the file is publicly accessible, but it's fine because it only contains anonymous data. But then we also discussed removing the "Email IPN report" in favour of this new logging system.. but if we integrate IPN reports then that would contain data like email addresses.

**Comment by mindctrl** [comment]
I wouldn't want us to log any personal data to a publicly accessible file. Given that, I think the IPN option should stay the same.

**Comment by nosegraze** [comment]
Updated PR:

- Added more logs.
- Also added more actions in the gateways for specific events. A lot of the existing actions were gateway-specific, i.e. `rcp_authorizenet_silent_post_payment`. Now they all share one like `rcp_gateway_payment_processed`.
- Removed old "Logs" menu item and deprecated WP_Logging class. I figured maybe we should keep the actual class there for a while in case people are using it in custom gateways?

I'll add logs for admin actions once #1101 is merged.

Here's a sample log for a single new registration:

```
2017-4-14 14:12:51 - Started new registration for subscription #15 via stripe.
2017-4-14 14:12:51 - Registration for user #1 sent to gateway. Level ID: 15; Initial Amount: 50; Recurring Amount: 50; Auto Renew: false
2017-4-14 14:12:55 - New payment inserted. ID: 106; User ID: 1; Amount: 50; Subscription: 1 Year Plan; Status: complete
2017-4-14 14:12:58 - Payment Received email sent to user #1.
2017-4-14 14:12:58 - New payment #106 inserted for member #1 via Stripe gateway.
2017-4-14 14:12:58 - Updating recurring status for member #1. Previous: false; New: false
2017-4-14 14:13:00 - Active email sent to user #1.
2017-4-14 14:13:02 - Active email sent to admin(s).
```

**Comment by mindctrl** [comment]
Adjustments to the constructor look good. I was able to whip up a quick test and do a custom log.

---

Seeing a couple log issues. One is really a symptom of a different issue, but worth mentioning.

Seeing duplicate entries when a new sub starts. That appears to be due to the init hook firing more than once, something we should probably address elsewhere, but maybe there's something we can do here. Maybe not.

```
2017-5-04 19:23:33 - Started new registration for subscription #2 via stripe.
2017-5-04 19:23:35 - Started new registration for subscription #2 via stripe.
```

Another is a missing transaction ID when a duplicate payment event fires.
```
2017-5-04 19:23:42 - A duplicate payment was detected for user #871. Check to make sure both payments weren't recorded. Transaction ID: 
```
Probably just need to pull it from the `$payment_event` object instead of the `$payment_data` array.

**Comment by mindctrl** [comment]
re: missing Transaction ID in Stripe log. I think we want `$payment_event->charge` in the `rcp_ipn_duplicate_payment` hook.

`do_action( 'rcp_ipn_duplicate_payment', $payment_event->charge, $member, $this );`

That gives the same transaction ID that is stored with the original payment.

**Comment by nosegraze** [comment]
I think it could either be `$payment_event->id` or `$payment_event->charge` depending on whether the event was `charge.succeeded` or `invoice.payment_succeeded`.

Based on the code though, it seems like `$payment_data['transaction_id']` should be set unless the event wasn't either of the two I mentioned. How did you test this?

**Comment by nosegraze** [comment]
Regarding the `Starting new registration` entry running twice, I think that's just because it runs once when checking for errors (when `$is_ajax` is `true`), then again after errors have been checked to process the actual registration (at which point `$is_ajax` is `false`).

**Comment by mindctrl** [comment]
It happens during auto-renew signups. Stripe sends both a `charge.succeeded` and `invoice.payment_succeeded`.

Looking back at the webhook data I have in ngrok, `invoice.payment_succeeded` came in first and RCP responded with `duplicate payment found`. Sounds like there's an issue in that code maybe. In either case, if the log says Transaction ID I think that's what we should use. If we want to log the invoice ID we can do that too/instead, we just need to label it that way.

I'll do some more testing today.

**Comment by mindctrl** [comment]
re: duplicate payment log missing transaction ID...

When the `invoice.payment_succeeded` webhook comes in on a auto-renewing subscription that was just created (no trials, no discounts, no negative signup fees - just a normal signup), the logic in the Stripe gateway is such that `rcp_ipn_duplicate_payment` runs even when it's not a duplicate payment. 

That's because `$payment_data['transaction_id']` is empty [here](https://github.com/restrictcontentpro/restrict-content-pro/blob/2.8.5/includes/gateways/class-rcp-payment-gateway-stripe.php#L537), and it flows to the `else` statement where `rcp_ipn_duplicate_payment` runs.

`$payment_data['transaction_id']` is empty because `$payment_event->charge` is *not* empty [here](https://github.com/restrictcontentpro/restrict-content-pro/blob/2.8.5/includes/gateways/class-rcp-payment-gateway-stripe.php#L521), so the transaction_id never gets set.

This issue existed before this PR, it's just that it's more apparent here because virtually every auto-renew subscription is going to log a duplicate payment entry.

**Comment by nosegraze** [comment]
Thanks for tracking that down. It makes sense now.

As a quick fix we could turn [that `else`](https://github.com/restrictcontentpro/restrict-content-pro/blob/2.8.5/includes/gateways/class-rcp-payment-gateway-stripe.php#L568) into an `elseif` like:

```php
elseif ( ! empty( $payment_data['transaction_id'] ) && $rcp_payments->payment_exists( $payment_data['transaction_id'] ) )
```

**Comment by mindctrl** [comment]
I think so too. Let's do it in a separate issue after merging this so we can have a clearer record of it.

**Comment by nosegraze** [comment]
👍

**Comment by mindctrl** [comment]
@nosegraze anything else you want to do here? If not, I'll give it one more spin and merge.

**Comment by nosegraze** [comment]
The only other thing I was going to do was add some notes when admin actions are performed, but I'd rather do that after https://github.com/restrictcontentpro/restrict-content-pro/issues/1101 is merged.

**Comment by nosegraze** [comment]
This is ready for a final look over now!

**Comment by mindctrl** [comment]
I made some comments on the PR, but I'm going to go ahead and update those, do some more testing, and merge.

**Comment by nosegraze** [comment]
I'm fine with those changes. :) Thanks!
